feat(android): migrate CloudPush to use newest Firebase APIs

jquick-axway opened the pull request:
**JIRA:**
- https://jira.appcelerator.org/browse/MOD-2634
- https://jira.appcelerator.org/browse/NATIVESDK-352
- https://jira.appcelerator.org/browse/TIMOB-27972

**Summary:**
- Updated `ti.cloudpush` module to use Firebase/FCM instead of GCM.
- Added new event `deviceTokenExpired`.
- Module correctly unregisters previous token when last token expires.
  * Caused by a breaking-change on Google's end due to our usage of deprecated API.

**Breaking-Change:**
Requires you to add `google-services.json` to project's `./platform/android` folder. Will now log an error at runtime if this file was missing when calling module's `retrieveDeviceToken()` method.

**Test:**
1. Create a Titanium app with Appc Platform Services (APS) enabled.
2. Set up the app for Cloud Messaging at the [Firebase Console](https://console.firebase.google.com/) website.
3. Copy the Firebase "Server Key" and "Sender ID" to the APS Cloud Backend for both "Development" and "Production".
4. Build and run for Android.
5. Verify that you see the following error in the log.
`CloudPush requires a Firebase "google-services.json" file...`
6. Download `google-services.json` from [Firebase Console](https://console.firebase.google.com/) to: `./platform/android`
7. Build and run on Android.
8. Verify that you see the following log message.
`Successfully subscribed for push notifications.`
9. Post a notification to the app via our website.
10. Verify that the app receives the notification.
11. Uninstall the app.
12. Re-install the app.
13. When you run the app, verify that you see the following log message.
`Previous CloudPush token has expired. Acquiring a new token.`
15. Send a notification to the app via our website.
16. Verify only 1 notification is posted to the app. (You should NOT receive 2 notifications.)
17. On our website, verify only 1 device is listed after posting the notification. (Note that before you sent the notification, you would see 2 devices, but posting the notification causes the old token to be auto-deleted.)

```javascript
const Cloud = require('ti.cloud');
const CloudPush = require('ti.cloudpush');

// Show notifications in status bar while app is backgrounded.
CloudPush.showTrayNotification = true;

// Show notifications in status bar while app is in the foreground.
CloudPush.showTrayNotificationsWhenFocused = true;

// Fetches app's push notification token and registers it with the cloud if changed.
function registerForPushNotifications() {
	CloudPush.retrieveDeviceToken({
		success: function (e) {
			// We've obtains a token. Do not continue if already subscribed it to the cloud.
			const deviceToken = e.deviceToken;
			Ti.API.info('Device Token: ' + deviceToken);
			const lastDeviceToken = Ti.App.Properties.getString('androidCloudPushToken', '');
			if (lastDeviceToken === deviceToken) {
				return;
			}

			// Log in to the cloud.
			let deployType = 'production';
			if (Ti.App.deployType.toLowerCase() !== deployType) {
				deployType = 'development';
			}
			Cloud.Users.login({
				login: Ti.App.Properties.getString('acs-username-' + deployType),
				password: Ti.App.Properties.getString('acs-password-' + deployType),
			}, function (e) {
				// Do not continue if failed to log in.
				if (!e.success) {
					Ti.API.error('Failed to log in to the cloud. Reason: ' + e.message);
					return;
				}

				// Send token to cloud so that this app can receive push notifications.
				Cloud.PushNotifications.subscribeToken({
					channel: 'alert2',
					device_token: deviceToken,
					type: 'gcm',
				}, function (e) {
					if (e.success) {
						// Store subscribed token to preferences.
						Ti.App.Properties.setString('androidCloudPushToken', deviceToken);
						Ti.API.info('Successfully subscribed for push notifications.');
					} else {
						Ti.API.error('Failed to subscribe for push notifications. Reason: ' + e.message);
					}
				});
			});
		},
		error: function (e) {
			// We've failed to acquire a token. Can happen if:
			// - App is not correctly configured for Firebase Messaging.
			// - Push notifications is disabled under system settings.
			// - Google Play is no installed on device.
			Ti.API.error('Failed to register for CloudPush. Reason: ' + e.error);
		}
	});
}

// Always fetch token from system on app startup in case it changed since last startup.
registerForPushNotifications();

// Called for every push notification received by the app.
CloudPush.addEventListener('callback', function (e) {
	alert(e.payload);
});

// Called when last token has expired and has become invalid.
// We must obtain a new token in order for this app to keep receiving push notifications.
CloudPush.addEventListener('deviceTokenExpired', function (e) {
	Ti.API.info('Previous CloudPush token has expired. Acquiring a new token.');
//	registerForPushNotifications();
});

// Called when cold-starting app by tapping notification in status bar.
CloudPush.addEventListener('trayClickLaunchedApp', function (e) {
	Ti.API.info('Tray Click Launched App (app was not running)');
});

// Called when resuming app from background by tapping notification in status bar.
CloudPush.addEventListener('trayClickFocusedApp', function (e) {
	Ti.API.info('Tray Click Focused App (app was already running)');
});
```

jquick-axway committed:
docs(android): updated for CloudPush 7.1.0

build commented:
<!--
  0 failure: 
  0 warning: 
  3 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11814/1/artifact/dist/mobilesdk-9.1.0.v20200710192213-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :white_check_mark: All tests are passing
Nice one! All 6693 tests are passing.
(There are 698 skipped tests not included in that total)
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 891104ad90b65fc1982e177903bf05559d5ea4dc
</p>

jquick-axway committed:
feat(android): migrate CloudPush to Firebase

- Updated "ti.cloudpush" module to v7.1.0.
- Removed usage of GCM and replaced with Firebase/FCM.
- Added new event "deviceTokenExpired".

sgtcoolguy committed:
Merge branch 'master' into MOD-2634

lokeshchdhry commented:
FR Passed.

- Not having `google-services.json` file gives error `[ERROR] :  Failed to register for CloudPush. Reason: CloudPush requires a Firebase "google-services.json" file to be added to the  Titanium project's "./platform/android" folder to work.` as expected.

- Adding the json file we can successfully subscribe to push notifications.

- Uninstalling the app & reinstalling again multiple times gives `Previous CloudPush token has expired. Acquiring a new token.` after some time & no duplicate push notifications are seen.

- Using the old module, installing the app, uninstalling, installing app with new module, its seen that the old GCM tokens are automatically deleted after few hrs (2 hrs for me) & the push functions as expected.

- Push works as expected for production builds as well.


Studio Ver:       6.0.0.202005141803
SDK Ver:          9.0.1.GA, 9.0.2.GA & 9.1.0 local build
OS Ver:           10.15.5
Xcode Ver:        Xcode 11.5
Appc NPM:         5.0.0
Appc CLI:         8.0.0
Daemon Ver:       3.2.0
Ti CLI Ver:       5.2.2
Alloy Ver:        1.14.6
Node Ver:         12.13.1
NPM Ver:          6.12.1
Java Ver:         11.0.6
Android Devices:  ⇨ google Pixel (Android 10)
CloudPush Module: 7.1.0

sgtcoolguy commented:
manually merged to master via rebase (after altering commit message to reference tickets)
